### PR TITLE
Editing period mark

### LINKED ISSUE #53: Edit period mark
Now, when you press the "mark" button, you must separately enter the start date and length of the period.
You need to do this: after pressing the "mark" button, a calendar appears, showing the expected days of period. The user can accept or edit them.

### PR DESCRIPTION
**Pull request by IraSoro**
Closed #53 

Previously, the user separately chose the start date and the length of the menstruation. This is incomprehensible.
Now the user is invited to mark the days of menstruation himself. Also, these dates are already calculated. The user can agree with these calculations or edit them.
There is one drawback. The user can enter dates not in a range. But they are processed as a range. Also, there is no restriction on future dans.

### CODE REVIEW

File: src/modals/MarkModal.tsx [hunk 1]
```diff
@@ -1,127 +1,83 @@
...
+
+    function nextPeriodDays(): string[] {
+        let periodDates = [format(nowDate, 'yyyy-MM-dd')];
+        let lengthOfNextPeriod = lengthOfPeriod;
+        if (lengthOfNextPeriod === 0) {
+            const averageLengthOfPeriod = 5;
+            lengthOfNextPeriod = averageLengthOfPeriod;
+        }
+
+        for (let i = 1; i < lengthOfNextPeriod; ++i) {
```
Comment (imblowfish) [inline src/modals/MarkModal.tsx:59]: You can rewrite this function to something like that:

```js
const periodDates: string[] = [];

for (let day = 0; day < lengthOfPeriod || 5; day++) {
  const periodDay = new Date(nowDate);
  periodDay.setHours(0, 0, 0, 0);
  periodDay.setDate(periodDay.getDate() + day);

  periodDates.push(format(periodDay, "yyyy-MM-dd");
}

return periodDates;
```

File: src/modals/MarkModal.tsx [hunk 2]
```diff
@@ -1,127 +1,83 @@
...
+            const nextPeriodDay: Date = new Date();
+            nextPeriodDay.setHours(0, 0, 0, 0);
+            nextPeriodDay.setDate(nextPeriodDay.getDate() + i);
+            periodDates.push(format(nextPeriodDay, 'yyyy-MM-dd'));
+        }
+
+        return periodDates;
+    }
+
+    const isEnabled = (dateString: string) => {
```
Comment (imblowfish) [inline src/modals/MarkModal.tsx:69]: `isEnabled` says nothing about this function here. Maybe rename it to something more telling?

File: src/modals/MarkModal.tsx [hunk 3]
```diff
@@ -137,38 +93,99 @@ const MarkModal = (props: PropsMarkModal) => {
...
-                    </IonCardContent>
-                </IonContent>
-            </IonModal>
+                <IonDatetime
+                    ref={datetimeRef}
+                    color="dark-basic"
+                    presentation="date"
+                    locale="en-GB"
+                    size="cover"
+                    preferWheel={false}
```
Comment (imblowfish) [inline src/modals/MarkModal.tsx:102]: Here and below you don't need to write `false` or `true` explicitly. You can just write `preferWheel` without `={false}`

File: src/pages/TabHome.tsx [hunk 4]
```diff
@@ -83,7 +83,9 @@ function useDaysBeforePeriod(): DaysBeforePeriod {
 
   const dateOfFinish = new Date(startDate);
   dateOfFinish.setDate(dateOfFinish.getDate() + Number(cycleLength));
+  dateOfFinish.setHours(0, 0, 0, 0);
```
Comment (imblowfish) [inline src/pages/TabHome.tsx:86]: As I can see we don't use hours/minutes/seconds in any place

Maybe just call `setHours` once when we save data to the storage/Context and remove all unnecessary `setHours`?

**Review by imblowfish — CHANGES_REQUESTED** [review]

File: src/modals/MarkModal.tsx [hunk 5]
```diff
@@ -1,127 +1,83 @@
...
-                onClick={() => {
-                    props.closeModal();
-                    props.setDate("");
-                }}>
-                Cancel
-            </IonButton>
-        </>
-    );
-};
+import { format } from 'date-fns'
```
[the remark this answers is not in the document]
Comment (IraSoro) [reply src/modals/MarkModal.tsx:15]: Fixed

File: src/modals/MarkModal.tsx [hunk 6]
```diff
@@ -137,38 +93,99 @@ const MarkModal = (props: PropsMarkModal) => {
...
+                    value={nextPeriodDays()}
+
+                    titleSelectedDatesFormatter={((selectedDates: string[]) => {
+                        if (selectedDates.length === 0) {
+                            return "select date range";
+                        }
+
+                        selectedDates.sort();
+                        const startPeriod = new Date(selectedDates[0]);
+                        const finishPeriod = new Date(selectedDates[selectedDates.length - 1]);
```
[the remark this answers is not in the document]
Comment (IraSoro) [reply src/modals/MarkModal.tsx:117]: Fixed

File: src/modals/MarkModal.tsx [hunk 7]
```diff
@@ -1,127 +1,83 @@
...
+    nowDate.setHours(0, 0, 0, 0);
+
+    const nextCycleFinish: Date = new Date();
+    nextCycleFinish.setDate(nowDate.getDate() + lengthOfPeriod);
+    nextCycleFinish.setHours(0, 0, 0, 0);
+
+    function isLastPeriodDay(date: Date) {
+        date.setHours(0, 0, 0, 0);
+
+        for (let i = 0; i < cycles.length; ++i) {
```
[the remark this answers is not in the document]
Comment (IraSoro) [reply src/modals/MarkModal.tsx:38]: Fixed

File: src/modals/MarkModal.tsx [hunk 8]
```diff
@@ -1,127 +1,83 @@
...
+        };
+
+        return false;
+    }
+
+    function nextPeriodDays(): string[] {
+        let periodDates = [format(nowDate, 'yyyy-MM-dd')];
+        let lengthOfNextPeriod = lengthOfPeriod;
+        if (lengthOfNextPeriod === 0) {
+            const averageLengthOfPeriod = 5;
```
[the remark this answers is not in the document]
Comment (IraSoro) [reply src/modals/MarkModal.tsx:55]: Fixed

File: src/modals/MarkModal.tsx [hunk 9]
```diff
@@ -1,127 +1,83 @@
...
+                return true;
+            }
+        };
+
+        return false;
+    }
+
+    function nextPeriodDays(): string[] {
+        let periodDates = [format(nowDate, 'yyyy-MM-dd')];
+        let lengthOfNextPeriod = lengthOfPeriod;
```
[the remark this answers is not in the document]
Comment (IraSoro) [reply src/modals/MarkModal.tsx:53]: And fixed

File: src/modals/MarkModal.tsx (hunk 1 above)
Comment (IraSoro) [reply src/modals/MarkModal.tsx:59]: Fixed

File: src/modals/MarkModal.tsx (hunk 2 above)
Comment (IraSoro) [reply src/modals/MarkModal.tsx:69]: Yes, I'm bad at it.
Fixed

File: src/modals/MarkModal.tsx (hunk 3 above)
Comment (IraSoro) [reply src/modals/MarkModal.tsx:102]: Ok!

**Review by imblowfish — APPROVED** [review]

**Pull request merged**